### PR TITLE
Hw 34 solution

### PR DESCRIPTION
**Pull request by ipantazi**
Added `Book` entity, `DataProcessingException` class, `BookRepository` interface with methods and `BookRepositoryImpl` class.
Added `BookService` interface with methods and `BookServiceImpl` class.
Modified `Application` class and file `application.properties`.
Created a new folder `resources` in the `src/test` directory
Create a new file `application.properties` in the `src/test/resources` folder.
Added dependencies to the `pom.xml` file.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -54,6 +54,16 @@
...
             <scope>runtime</scope>
         </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>commons-dbcp</groupId>
+            <artifactId>commons-dbcp</artifactId>
```
Comment (Elena-Bruyako) [inline pom.xml]: Why do you need such dependency?

File: src/main/java/mate/academy/onlinebookstore/Application.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    private BookService bookService;
+
+    public static void main(String[] args) {
+        SpringApplication.run(Application.class, args);
+    }
+
+    @Bean
+    public CommandLineRunner commandLineRunner() {
+        return args -> {
+            Book macbeth = new Book();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/Application.java]:
```suggestion
            Book book = new Book();
```

File: src/main/java/mate/academy/onlinebookstore/Application.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            macbeth.setTitle("Macbeth");
+            macbeth.setAuthor("William Shakespeare");
+            macbeth.setIsbn("978-0091547714");
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/Application.java]: set up all fields

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Entity
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]: Don't use `@Data` for entities, find and read articles about problems
Lombok has a lot of other annotations that replace boilerplate code
Don't forget to add Table annotation, to specify the name of table according to the naming convention

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+public class Book {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String title;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java:24]: add required constraints for fields

File: src/main/java/mate/academy/onlinebookstore/repository/BookRepository.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.onlinebookstore.repository;
+
+import java.util.List;
+import mate.academy.onlinebookstore.model.Book;
+
+public interface BookRepository {
+    Book save(Book book);
+
+    List<Book> findAll();
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/repository/BookRepository.java]:
```suggestion
```

File: src/main/java/mate/academy/onlinebookstore/repository/BookRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.EntityManagerFactory;
+import jakarta.persistence.EntityTransaction;
+import java.util.List;
+import mate.academy.onlinebookstore.exception.DataProcessingException;
+import mate.academy.onlinebookstore.model.Book;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/repository/BookRepositoryImpl.java:12]:
```suggestion
@Repository
@RequiredArgsConstructor
```

File: src/main/java/mate/academy/onlinebookstore/repository/BookRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+public class BookRepositoryImpl implements BookRepository {
+    private final EntityManagerFactory entityManagerFactory;
+
+    @Autowired
+    public BookRepositoryImpl(EntityManagerFactory entityManagerFactory) {
+        this.entityManagerFactory = entityManagerFactory;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/repository/BookRepositoryImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
+package mate.academy.onlinebookstore.service;
+
+import java.util.List;
+import mate.academy.onlinebookstore.model.Book;
+import mate.academy.onlinebookstore.repository.BookRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java:9]:
```suggestion
@Service
@RequiredArgsConstructor
```

File: src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+public class BookServiceImpl implements BookService {
+    private final BookRepository bookRepository;
+
+    @Autowired
+    public BookServiceImpl(BookRepository bookRepository) {
+        this.bookRepository = bookRepository;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/service/BookServiceImpl.java]:
```suggestion
```

File: src/main/resources/application.properties [hunk 11]
```diff
@@ -1 +1,9 @@
 spring.application.name=Online-Book-store
+
+spring.datasource.url=jdbc:mysql://localhost:3306/my_db?serverTimezone=UTC
```
Comment (Elena-Bruyako) [inline src/main/resources/application.properties]: it's better to rename schema according to your project (don't be so abstract)

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    private BigDecimal price;
+
+    @Column(name = "description", nullable = false)
+    private String description;
+
+    @Column(name = "cover_image", length = 2048, nullable = false)
+    private String coverImage;
+
+    @Override
+    public final boolean equals(Object o) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]: remove

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+                && (Objects.equals(title, book.title))
+                && (Objects.equals(author, book.author))
+                && (Objects.equals(isbn, book.isbn))
+                && (Objects.equals(price, book.price))
+                && (Objects.equals(description, book.description))
+                && (Objects.equals(coverImage, book.coverImage));
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]: remove

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+import lombok.ToString;
+
+@Entity
+@Table(name = "books")
+@Getter
+@Setter
+@Builder
+@ToString
+@NoArgsConstructor
+@AllArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]:
```suggestion
```

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import lombok.ToString;
+
+@Entity
+@Table(name = "books")
+@Getter
+@Setter
+@Builder
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]: do you use builder somewhere? If no - remove

File: src/main/java/mate/academy/onlinebookstore/model/Book.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+@ToString
+@NoArgsConstructor
+@AllArgsConstructor
+public class Book {
+    @Column(name = "id", nullable = false)
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(name = "title", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/onlinebookstore/model/Book.java]:
```suggestion
    @Column(nullable = false)
```

for all fields

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request closed without merging**